src/providers/MeetingProvider/MeetingManager.ts

```diff
@@ -101,6 +101,18 @@ export class MeetingManager implements AudioVideoObserver {
 
   simulcastEnabled: boolean = false;
 
+  skipAudioInputProviderInit: (() => void) | null = null;
+
+  skipAudioOutputProviderInit: (() => void) | null = null;
+
+  skipVideoInputProviderInit: (() => void) | null = null;
+
+  invokeAudioInputProviderInit: (() => void) | null = null;
+
+  invokeAudioOutputProviderInit: (() => void) | null = null;
+
+  invokeVideoInputProviderInit: (() => void) | null = null;
+
```

Comment:
@richnew10 These functions is designed to receive the callbacks from those `DeviceProvider`s to control the device permission request. The logic is quite straightforward now, but it looks bad. Do you think I need to avoid them or improve them to subscribe pattern? Or do you have any better idea?

Reply:
I don't see why those providers need to know anything about the device label hook. They don't use labels, right?

Reply:
I'm sure I've written this before, but: there are only two things that trigger a permissions request — choosing a device and the device label hook. The only thing we need to care about here is the device label hook, and that's controlled by the MeetingManager. The only thing you need to do in this PR is provide a way to configure the MeetingManager to control the device label hook.

Reply:
@richnew10 For example, there is a `initAudioInput()` function in `AudioInputProvider`, and it's called after rendering the `AudioInputProvider`. `initAudioInput` function calls `audioVideo.listAudioInputDevices()`, which will call `deviceLabelTrigger` in JS SDK. 

So I have to block/skip the `initAudioInput()` calling, and invoke it when user wants to grant permission to browser.

There could be another way, we just mock an empty `deviceLabelTrigger` funcion via `setupDeviceLabelTrigger`. So even the `initAudioInput()` is called, it won't trigger any permission request. I'm not sure how to mock an empty `deviceLabelTrigger`, cause the `getUserMedia` requires either `audio` or `video` to be `true`. I would appreciate it if you can teach me how to do this.

Reply:
Yes, that's exactly the way I suggest you do it — 

```
defaultDeviceController.setDeviceLabelTrigger(() => Promise.resolve(new MediaStream()));
```

should be sufficient if the user doesn't want a permissions dialog to be triggered. That will cause device labels to be missing if the user hasn't granted permissions.

Reply:
@richnew10 Cool, now I know what to do! Thank you so much!

Reply:
@richnew10 I tried to improve the code as you suggested. It works well on suppressing device permission request. But there is new problem. Although we set an empty `deviceLabelTrigger` to JS SDK, but we still call `audioVideo.listDevicesOfKind` for each kind of device in `DeviceProvider`. The devices info with empty label will be cached in `deviceInfoCache`. And when we want to invoke the `DeviceProvider` to trigger device permission request, we have to run `audioVideo.listDevicesOfKind` for the desired deviceKind again. At this time the device info with empty label in the `deviceInfoCache` will be directly returned. The `deviceLabelTrigger` won't be called, so there is no device permission request and the devices info won't be updated. How could we solve this problem? 

Potential Solution:
* somehow reset `deviceInfoCache`
* somehow force `deviceLabelTrigger` being called

Since this is a change to component SDK, I think we should avoid the change to JS SDK.

```ts
private async listDevicesOfKind(deviceKind: string): Promise<MediaDeviceInfo[]> {
    if (this.deviceInfoCache === null || !this.isWatchingForDeviceChanges()) {
    // This function is skipped in the second calling
      await this.updateDeviceInfoCacheFromBrowser();
    }
    return this.listCachedDevicesOfKind(deviceKind);
  }

  private async updateDeviceInfoCacheFromBrowser(): Promise<void> {
    const doesNotHaveAccessToMediaDevices = typeof MediaDeviceInfo === 'undefined';
    if (doesNotHaveAccessToMediaDevices) {
      this.deviceInfoCache = [];
      return;
    }

    let devices = await navigator.mediaDevices.enumerateDevices();

    let hasDeviceLabels = true;
    for (const device of devices) {
      if (!device.label) {
        hasDeviceLabels = false;
        break;
      }
    }
    if (!hasDeviceLabels) {
      try {
        this.logger.info('attempting to trigger media device labels since they are hidden');
        const triggerStream = await this.deviceLabelTrigger();
        devices = await navigator.mediaDevices.enumerateDevices();
        for (const track of triggerStream.getTracks()) {
          track.stop();
        }
      } catch (err) {
        this.logger.info('unable to get media device labels');
      }
    }
    this.deviceInfoCache = devices;
  }
```

Reply:
If the builder never calls `chooseAudioOutputDevice` or selects an input device or adds a device change observer then `onDeviceChangeCallback` will not be set, and so `isWatchingForDeviceChanges` will be false, and so the cache should be re-updated, no? So the issue is that they've selected a media stream but we've still decided to subscribe to device change notifications.

Regardless, I think this is worth fixing in the SDK, because it affects customers who don't use the component library:

```typescript
ddc.setDeviceLabelTrigger(() => Promise.resolve(new MediaStream()));
await ddc.listAudioInputDevices();
ddc.setDeviceLabelTrigger(() => navigator.mediaDevices.getUserMedia({audio: true}));
await ddc.listAudioInputDevices();  // < returns stale data, and no way to change it
```

If there are no device labels and no cache, we perhaps shouldn't set `deviceInfoCache` at the end of `updateDeviceInfoCacheFromBrowser`.

If the builder changes the trigger, we should perhaps throw away the cache.

And it also makes sense to allow the builder to discard the cache manually without recreating the device controller, or to force an update by adding a `force` parameter to the list functions.

Reply:
@xuesichao Agreed with Richard here. Let add an API to refresh device cache in JS SDK.

Reply:
Ok I get it! Let me do this. Thanks guys!

Reply:
@richnew10 I updated the PR with my current fix. The `invokeDeviceProvider(deviceLabels)` works well right now. The updated PR should pass the GitHub check successfully after next JS SDK release.

The last part left is I need to handle the logic when users pass a `deviceLabels` of `DeviceLabelTrigger` type to `meetingManager.join(deviceLabels)`.  I think the main difference should be the constraints in `getUserMedia()`.

```js
{ audio: true, video: true }
```

Since we already cover all the cases in`DeviceLabels` type, I'm curious how should we handle this case?

I mean if user pass a customized `deviceLabelTrigger` like `meetingManager.join(deviceLabelsTrigger)`, how should we control the logic of `DeviceProvider`. The invoke logic can be controlled by the users themselves. But how could we know which `DeviceProvider` should be delayed?